### ISSUE TITLE
Distant Supervision for Relation Extraction via Piecewise Convolutional Neural Networks (EMNLP-2015/09)

### ISSUE THREAD

**Issue by BrambleXu** [BrambleXu/knowledge-graph-learning#14]
# 一句话总结

把Distant Supervision问题当做multi-instance问题就解决错误标注的情况。并引入PCNN提取segment更细节的information。

# 资源

- [code](https://github.com/ShomyLiu/pytorch-pcnn)

# 概要

## 要解决的问题

对于RE任务，distant supervision有两个问题。

1. knowledge base是以启发式的方式与文本对齐的，对齐的结果则作为标注数据。然而，这种启发式对齐可能产生错误的标签。
2. 针对distant supervision，以前的一些统计模型经常使用一些特殊的特征（ad hoc features），抽取这些特征带来的noise，也会导致最终效果变差。

## 解决方法

1. 针对第一个问题，distant supervised relation extraction被当做一个multi-instance label问题来进行处理
2. 针对第二个问题，不采用特征工程的方式来提取特征，而是用CNN+piecewise max pooling的方式来抽取特征

## 1. 简介

针对RE，一个挑战是如何产生足够的训练数据。而distant supervision被提出来了。

- distant supervision：假设在一个knowledge base里，如果两个entity有一个relation，那么我们认为包含这两个entity的所有句子，都包含这个relation。

![](https://ws1.sinaimg.cn/large/006tKfTcgy1g0ewqaa83ej30gl09djrq.jpg)

但是这种方式有两个很大的缺点

- 从上面的例子可以看到，这种方式抽取出的句子不一定能正确表示两个entity的relation。因为虽然两个entity会在同一个句子里，但并不代表这个句子有两个entity的关系。两个entity可能只是分享一个topic罢了
- 以前的一些方法，都是使用监督式模型来设计特征，然后通过distant supervision来获取labeled data. 这些通过NLP tools提取出来的特征，可能会传递和累计错误。下面是，某数据集句子长度的分布。一半以上的句子长度都超过了40单词。而根据研究，syntactic parsing的精度会随着句子长度的增加显著下降。所以传统的特征抽取方式抽出的feature是有很多错误的。

![image-20190222094754775](https://ws4.sinaimg.cn/large/006tKfTcgy1g0ewvhgyooj30gi0btq3v.jpg)

1. 针对第一个问题，distant supervised relation extraction被当做一个multi-instance label问题来进行处理。In multi-instance problem, the training set consists of many bags, and each contains many instances. The labels of the bags are known; however, the labels of the instances in the bags are unknown. We design an objective function at the bag level. In the learning process, the uncertainty of instance labels can be taken into account; this alleviates the wrong label problem. （对于多实例问题，训练数据包含很多bag，每一个bag包含很多实例。每个bag的label是知道的，但是bag里实例的label是不知道的。为了解决这个问题，我们在bag level设计了一个目标函数。通过这个目标函数，在训练过程中，实例的不确定性会被考虑，这个会缓解错误的标签问题）
2. 针对第二个问题，不采用特征工程的方式来提取特征，而是用CNN+piecewise max pooling的方式来抽取特征。这个主要是用了Zeng 2014 #13 的方法，文本的PCNN相当于是之前的一个扩展。
   - 通过CNN的方式有一个缺点，就是无法捕捉两个entity之间的structrual information。为了解决这个问题，大部分工作都是通过manually crafted features来给structrual information建模的。这些工作考虑internal context和external context。根据给定的两个entity，一个句子通常可以分割为3部分。internal context包含两个entity里包含的character，而external context包含两个entity周围的characers。
   - 很显然，single max pooling无法捕捉这些信息。因此，我们把一个句子分成3个segment，然后用CNN对3个segment分别做max pooling。这样的话piecewise max pooling procedure会针对每个segment返回一个最大值，而不是针对整个sentence返回一个最大值。

# 3 Methodology

![image-20190222101004022](https://ws1.sinaimg.cn/large/006tKfTcgy1g0exij5lt7j30qa0cm41b.jpg)

Figure 3 shows our neural network architecture for distant supervised relation extraction. It illustrates the procedure that handles one instance of a bag. 这个过程包含四个部分: Vector Representation, Convolution, Piecewise Max Pooling and Softmax Output. We describe these parts in detail below.

### 3.1.1 Word Embeddings

与之前面的#13 使用的embedding不同，这里用了word2vec

### 3.1.2 Position Embeddings

![image-20190222101428091](https://ws2.sinaimg.cn/large/006tKfTcgy1g0exn3xqlqj30eu076jtb.jpg)

为什么选择son作为中点呢？是因为son、表示关系吗？如果句子里没有这个表示关系的词的话，怎么办？

### 3.2 Convolution

卷积的说明

Convolution is an operation between a vector of weights, w, and a vector of inputs that is treated as a sequence q

## 3.3 Piecewise Max Pooling

![image-20190222102140492](https://ws1.sinaimg.cn/large/006tKfTcgy1g0exulpl9xj30f40egtd3.jpg)

将卷积的结果做Piecewise Max Pooling，根据两个entity的位置分别做 Max Pooling。具体可以看图容易理解。每一个filter的结果通过Piecewise Max Pooling得到一个3位的向量，然后把这个向量与其他filter通过Piecewise Max Pooling的结果结合起来。

![image-20190222102316492](https://ws2.sinaimg.cn/large/006tKfTcgy1g0exw9nyd4j30fh070q4g.jpg)

最后得到的向量维度与原来的句子长度无关。

## 3.4 Softmax Output

![image-20190222102349917](https://ws4.sinaimg.cn/large/006tKfTcgy1g0exwuknv9j30f107rjt8.jpg)

最后一步通过softmax输出的结果$o_i$表是对每个$relaiton_i$的概率。

![image-20190222102635751](https://ws3.sinaimg.cn/large/006tKfTcgy1g0exzq5e2ej30fr096aco.jpg)

在倒数第二层用一个dropout来进行正则化。具体的思想就是用一个masking操作来屏蔽掉一些神经元。

![image-20190222102807654](https://ws1.sinaimg.cn/large/006tKfTcgy1g0ey1boeutj30f406pac3.jpg)

测试的时候不能使用dropout，即关闭masking。

## 3.5 Multi-instance Learning

![image-20190222104807007](https://ws3.sinaimg.cn/large/006tKfTcgy1g0eym4k2zxj30en06bgnq.jpg)

为了解决第一个问题，使用multi-instance learning for PCNN。还记得刚才说的吗？通过设计一个更有效的目标函数来考虑bag里的没有标签的实例。

![image-20190222105021312](https://ws1.sinaimg.cn/large/006tKfTcgy1g0eyosusfij30f50h8q7c.jpg)

因为之前的目标函数是针对instance的，所以这里要把目标函数改为针对bag的。等等。

整个算法流程为：

![image-20190222105257427](https://ws3.sinaimg.cn/large/006tKfTcgy1g0eyr5iawtj30fs0bjdis.jpg)

传统的反向传递是根据所有训练实例的，但是multi-instance learning则是基于bag的。因为一个bag里有标注错误的句子，所以我们基于bag训练的结果，自然也把这种错误的信息学了进去。

训练好PCNN之后，用于预测的时候，只有当一个bag里起码有一个实例被标记为正标签的时候，这个bag才会被标记为正标签。

# 4 Experiments

## 4.1 Dataset and Evaluation Metrics

We evaluate our method on a widely used dataset 4 that was developed by (Riedel et al., 2010) and has also been used by (Hoffmann et al., 2011; Surdeanu et al., 2012). This dataset was generated by aligning Freebase relations with the NYT corpus, with sentences from the years 2005-2006 used as the training corpus and sentences from 2007 used as the testing corpus.

### 4.2.1 Pre-trained Word Embeddings

word2vec，每个单词50维。

### 4.2.2 Parameter Settings

![image-20190222105919215](https://ws4.sinaimg.cn/large/006tKfTcgy1g0eyxruw67j30ma03v0tk.jpg)

### 4.3.1 Held-out Evaluation

![image-20190222110440638](https://ws3.sinaimg.cn/large/006tKfTcgy1g0ez3ctmaoj30im07tjv1.jpg)

Half of the Freebase relations are used for testing.

MIML is a multi-instance multilabel model that was proposed by (Surdeanu et al., 2012).PCNNs+MIL denotes our method

![image-20190222112352995](https://ws3.sinaimg.cn/large/006tKfTcgy1g0eznbqro4j30e706zq4a.jpg)

### 4.3.2 Manual Evaluation

![image-20190222112402679](https://ws4.sinaimg.cn/large/006tKfTcgy1g0eznhz126j30dw0c6wgf.jpg)

Thus, the held-out evaluation suffers from false negatives in Freebase. We perform a manual evaluation to eliminate these problems. For the manual evaluation, we choose the entity pairs for which at least one participating entity is not present in Freebase as a candidate. This means that there is no overlap between the held-out and manual candidates.

## 4.4 Effect of Piecewise Max Pooling and Multi-instance Learning

![image-20190222112519603](https://ws1.sinaimg.cn/large/006tKfTcgy1g0ezouiwoij30fb0d6mzg.jpg)

Both CNNs+MIL and PCNNs+MIL outperform their counterparts CNNs and PCNNs, respectively, thereby demonstrating that incorporation of multi-instance learning into our neural network was successful in solving the wrong label problem.
